ERC2981 (Royalty Info) for Cairo

immrsd opened the pull request:
- Add support for `ERC2981` (RoyaltyInfo) to `ERC721` and `ERC1155`
- Support `uint` Cairo types in input fields
- Support constants in trait implementations

<img width="1309" alt="image" src="https://github.com/user-attachments/assets/6cbd499a-74f3-4e75-8d4c-3051c96d2382">

ericnordelo reviewed (approved):
The changes look great @immrsd. We should think of a way to group imports together, but that's out of the scope of this PR.

andrew-fleming commented on packages/core-cairo/src/utils/convert-strings.ts:
```diff
@@ -45,4 +45,50 @@ export function toFelt252(str: string, field: string): string {
+/**
+ * Validates a string value to be a valid uint and converts it to bigint
+ */
+export function toUint(str: string, field: string, type: UintType): bigint {
+  const isValidNumber = /^\d+$/.test(str);
+  if (!isValidNumber) {
+    throw new OptionsError({
+      [field]: 'Not a valid number'
+    });
+  }
+  const numValue = BigInt(str);
+  if (numValue < BigInt(0)) {
+    throw new OptionsError({
+      [field]: `${type} value cannot be negative`
+    });
```
Inserting negative values into the default royalty fraction and fee denominator fields returns the first options error `Not a valid number`. The regex only matches numbers so negative numbers with the `-` character will never make it to the second options error, no?

immrsd replied:
Indeed, this check has no effect, removed it

andrew-fleming reviewed (commented):
Nice work, @immrsd! Overall, the PR looks good to me. I left a few questions and comments

ericglau reviewed (commented):
Thanks @immrsd, this looks great! Just a few comments.

immrsd commented on packages/ui/src/cairo/inject-hyperlinks.ts:
```diff
@@ -33,7 +33,7 @@ function removeComponentName(libraryPathSegments: Array<string>) {
   if (lastItem !== undefined && componentMappings[lastItem] !== undefined) {
     // Replace component name with the name of its .cairo file
-    libraryPathSegments.splice(-1, 1, componentMappings[lastItem]);
+    libraryPathSegments.splice(-1, 1, componentMappings[lastItem] as string);
```
It's needed (at least in the Typescript version I'm using). There's a check `componentMappings[lastItem] !== undefined`, but since the dictionary is mutable, the compiler can't guarantee that on the next line an item at the `lastItem` will still exist

<img width="928" alt="image" src="https://github.com/user-attachments/assets/7c523ce6-ec7f-437b-b5cc-a637f0eb0056">

immrsd commented on packages/ui/src/cairo/ERC1155Controls.svelte:
```diff
@@ -70,6 +73,8 @@
   </div>
 </section>
 
+<RoyaltyInfoSection bind:opts={opts.royaltyInfo} errors={errors} />
+
```
Agree, we should consider that

On the other hand, the Cairo control panels are still more concise that the Solidity ones and if we decide to go with that approach it may make sense to also apply it to the Solidity version

immrsd commented on packages/core-cairo/src/set-royalty-info.ts:
```diff
@@ -0,0 +1,123 @@
+
+export function setRoyaltyInfoIfNeeded(c: ContractBuilder, options: RoyaltyInfoOptions, access: Access): boolean {
+  if (!options.enabled) {
+    return false;
```
Makes sense, wanted to make the return value tell if the component was added or not, but it makes more sense to add it later if it's actually going to be used somewhere

immrsd commented on packages/core-cairo/src/set-royalty-info.ts:
```diff
@@ -0,0 +1,123 @@
+  enabled: true,
+  defaultRoyaltyFraction: '10',
+  feeDenominator: DEFAULT_FEE_DENOMINATOR.toString()
+};
```
I agree this is the desired approach, so I've added 3 different options. The only problem is that now instead of 800 generated contracts we have 2000. That shouldn't be a problem, but as a result Scarb compiler required more than 100 GB of RAM at some point during compilation. I've barely managed to compile all the contracts and it took more than 30 minutes

Scarb compiler is not yet well-optimized and this behaviour shouldn't be normal. We should look into it, it seems that some compilation steps that should be O(1) are in fact O(n). Although, I reckon it shouldn't be in the scope of this PR

<img width="487" alt="Screenshot 2024-11-29 at 09 10 36" src="https://github.com/user-attachments/assets/1a94f4c4-f5b3-4105-8b21-6dc219ec1cb5">

ericglau reviewed (approved):
It would be good to add some tests in `packages/core-cairo/src/erc721.test.ts` and `packages/core-cairo/src/erc1155.test.ts` so that their snapshots are included.

Can you also bump the version in `packages/core-cairo/package.json`?

The rest LGTM, thanks!

immrsd commented on packages/core-cairo/src/generate/erc1155.ts:
```diff
@@ -13,8 +14,9 @@ const blueprint = {
   updatableUri: booleans,
-  access: accessOptions,
   upgradeable: upgradeableOptions,
+  royaltyInfo: [royaltyInfoEnabled],
```
Addressed and added 3 different testing options

immrsd commented:
Updated the test project 

> It would be good to add some tests in `packages/core-cairo/src/erc721.test.ts` and `packages/core-cairo/src/erc1155.test.ts` so that their snapshots are included.
> 
> Can you also bump the version in `packages/core-cairo/package.json`?
> 
> The rest LGTM, thanks!

Addressed and added tests for royalty info feature for ERC721 and ERC1155

ericglau commented:
@SocketSecurity ignore npm/ws@7.4.6

Unrelated to this PR.
